test/unit/machine-learning/machine-learning-api-client.spec.ts

```diff
@@ -386,6 +415,153 @@ describe('MachineLearningApiClient', () => {
+    });
+
+    it('handles a running operation with no wait', () => {
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .resolves(utils.responseFrom(LOCKED_MODEL_RESPONSE));
+      stubs.push(stub);
+      return apiClient.handleOperation(OPERATION_NOT_DONE_RESPONSE)
+        .then((resp) => {
+          expect(resp).to.deep.equal(LOCKED_MODEL_RESPONSE);
+          expect(stub).to.have.been.calledOnce.and.calledWith({
+            method: 'GET',
+            url: `${BASE_URL}/projects/${PROJECT_ID}/models/${MODEL_ID}`,
+            headers: EXPECTED_HEADERS,
+          });
+        });
+    });
+
+    it('handles a running operation with wait', () => {
+      const stub = sinon.stub(HttpClient.prototype, 'send');
+      stub.onCall(0).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stub.onCall(1).resolves(utils.responseFrom(OPERATION_SUCCESS_RESPONSE));
+      stubs.push(stub);
+      return apiClient.handleOperation(OPERATION_NOT_DONE_RESPONSE, {
+        wait: true,
+        maxTimeMillis: 1000,
+        baseWaitMillis: 2,
+        maxWaitMillis: 5 })
+        .then((resp) => {
+          expect(resp).to.deep.equal(MODEL_RESPONSE);
+          expect(stub).to.have.been.calledTwice.and.calledWith({
+            method: 'GET',
+            url: `${BASE_URL}/projects/${PROJECT_NUMBER}/operations/${OPERATION_ID}`,
+            headers: EXPECTED_HEADERS,
+          });
+        });
+    });
+
+    it('handles a running operation with wait ending in error', () => {
+      const stub = sinon.stub(HttpClient.prototype, 'send');
+      stub.onCall(0).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stub.onCall(1).resolves(utils.responseFrom(OPERATION_ERROR_RESPONSE));
+      stubs.push(stub);
+      const expected = new FirebaseMachineLearningError('invalid-argument', STATUS_ERROR_MESSAGE);
+      return apiClient.handleOperation(OPERATION_NOT_DONE_RESPONSE, {
+        wait: true,
+        maxTimeMillis: 1000,
+        baseWaitMillis: 2,
+        maxWaitMillis: 5 })
+        .should.eventually.be.rejected.and.deep.include(expected)
+        .then(() => {
+          expect(stub).to.have.been.calledTwice.and.calledWith({
+            method: 'GET',
+            url: `${BASE_URL}/projects/${PROJECT_NUMBER}/operations/${OPERATION_ID}`,
+            headers: EXPECTED_HEADERS,
+          });
+        });
+    });
+
+    it('handles a running operation with wait ending in timeout', () => {
+      const stub = sinon.stub(HttpClient.prototype, 'send');
+      stub.onCall(0).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stub.onCall(1).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stub.onCall(2).resolves(utils.responseFrom(OPERATION_NOT_DONE_RESPONSE));
+      stubs.push(stub);
+      const expected = new Error('ExponentialBackoffPoller dealine exceeded - Master timeout reached');
+      return apiClient.handleOperation(OPERATION_NOT_DONE_RESPONSE, {
```

Comment:
Is this timing sensitive by any chance? I assume it waits 500ms before making the first call. Then makes a second call after 1s. At which point the maxTimeMillis has already passed so we get a timeout. Is that correct?

Reply:
Yes, that's correct. It should not be timing sensitive.